### PR TITLE
added optional optimization for einsum

### PR DESCRIPTION
**Pull request by mrader1248**
As I pointed out in [this issue](https://github.com/tensorflow/tensorflow/issues/19776), the `einsum` function does not have an argument `optimize` like its numpy counterpart. For this reason I implemented an optional depth-first search over pairwise contractions. The default behaviour is still the same as before, so no optimisations are performed, unless `optimize=True` or `optimize='dfspairs'` is set.

### CODE REVIEW

**Comment by mrader1248** [comment]
I signed it!

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @yifeif: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 16 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @yifeif: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 14 days with no activity and the `awaiting review` label has been applied.

**Comment by tensorflowbutler** [comment]
Nagging Reviewer @yifeif: You have been added as a reviewer to this pull request. Please add your review or reassign. It has been 29 days with no activity and the `awaiting review` label has been applied.

File: tensorflow/python/ops/special_math_ops.py [hunk 1]
```diff
@@ -134,64 +137,57 @@ def bessel_i1(x, name=None):
...
   which is written in a shorthand form inspired by the Einstein summation
   convention.  As an example, consider multiplying two matrices
   A and B to form a matrix C.  The elements of C are given by:
-
   ```
     C[i,k] = sum_j A[i,j] * B[j,k]
   ```
-
   The corresponding `equation` is:
-
```
Comment (martinwicke) [inline tensorflow/python/ops/special_math_ops.py]: Does this still render correctly without the empty lines?

@MarkDaoust do you know without testing?

**Review by martinwicke — CHANGES_REQUESTED** [review]
Wow this is quite impressive. Can you please add tests that verify

- the results using these optimizations are correct
- these optimizations are being used

The latter may be complicated, but maybe you can assert something about the structure of the resulting graph? If it's too onerous, we can live without -- but testing that this does what it should is mandatory.

File: tensorflow/python/ops/special_math_ops.py [hunk 2]
```diff
@@ -134,64 +137,57 @@ def bessel_i1(x, name=None):
...
   ```
     ij,jk->ik
   ```
-
   In general, the `equation` is obtained from the more familiar element-wise
   equation by
     1. removing variable names, brackets, and commas,
```
Comment (MarkDaoust) [inline tensorflow/python/ops/special_math_ops.py]: In fact, could you please add a blank like here, so this list renders correctly?

https://www.tensorflow.org/api_docs/python/tf/einsum

File: tensorflow/python/ops/special_math_ops.py (hunk 1 above)
Comment (MarkDaoust) [reply tensorflow/python/ops/special_math_ops.py]: >Does this still render correctly without the empty lines?

It won't, please leave them in.

**Comment by mrader1248** [comment]
>     * the results using these optimizations are correct

Would it be okay, if I just write tests for the functions `_einsum_optimize_dp` and `_einsum_optimize_greedy`, which I use internally to find the optimized contraction sequence?
 
>     * these optimizations are being used

I guess, to deduce this from the graph structure is extremely tedious. Would it be okay to check this via runtime comparisons?

File: tensorflow/python/ops/special_math_ops.py (hunk 1 above)
Comment (mrader1248) [reply tensorflow/python/ops/special_math_ops.py]: Thank you for pointing this out. Apparently, some copy/paste-step caused them to vanish..

File: tensorflow/python/ops/special_math_ops.py [hunk 3]
```diff
@@ -452,6 +485,188 @@ def _total_size(shape_values):
...
+              new_shape = tuple(
+                [d for d, f in zip(d1, m1) if f] + \
+                [d for d, f in zip(d2, m2) if f]
+              )
+
+              new_labels = \
+                ''.join([l for l, f in zip(l1, m1) if f]) + \
+                ''.join([l for l, f in zip(l2, m2) if f])
+              
+              contraction_cost = np.prod(d1) * \
```
Comment (MarkDaoust) [inline tensorflow/python/ops/special_math_ops.py]: What happens if one of the input shapes is unknown? 
Maybe I missed it, but I don't see a any check for [None] in the shapes.
Don't you need to fallback to "optimize=False" in that case?

Comment (mrader1248) [reply tensorflow/python/ops/special_math_ops.py]: Already now `einsum` just fails, if input shapes are not given. Therefore, also `optimize=False` won't work. Just try the following code:
```python
a = tf.placeholder(tf.float64)
b = tf.placeholder(tf.float64)
c = tf.einsum("ij,jk", a, b)
```
This results with:
```
ValueError: Cannot take the length of Shape with unknown rank.
```

Comment (MarkDaoust) [reply tensorflow/python/ops/special_math_ops.py]: I wouldn't expect that to work either.

The case I'm interested in is with known ranks, but unknown dimensions:

```
a = tf.placeholder(tf.float64, [10,None])
b = tf.placeholder(tf.float64, [None,7])
c = tf.einsum("ij,jk", a, b)
```
```
<tf.Tensor 'einsum_2/transpose_2:0' shape=(10, 7) dtype=float64>
```

For a real world example think of the transformer attention layers, where the sequence lengths are unknown at "compile" time. 

```
# query is [batch, sequence_s, depth]
# key is [batch, sequence_t, depth]

einsum('bsd,btd-->bst', query, key)
```

**Comment by rthadur** [comment]
@mrader1248 can you please resolve conflicts.

**Comment by rthadur** [comment]
> @mrader1248 can you please resolve conflicts.

@mrader1248  gentle ping

**Comment by mrader1248** [comment]
I'm sorry that it took me so long. The optimizations are now performed in the function `einsum_optimize` that is called from `einsum`.
I modified the method `run_test` in `EinsumTest` such that all test cases are tested with all optimization strategies. This ensures that results produced by different optimization strategies are equal and correct.
Furthermore, I added a test class `EinsumOptimizeTest` that has several test cases to make sure that optimizations are actually performed in the desired way.

**Comment by dgasmith** [comment]
You can likely use the built in NumPy contraction path optimization routines [here](https://docs.scipy.org/doc/numpy/reference/generated/numpy.einsum_path.html#numpy.einsum_path) or see more advanced routines [here](https://optimized-einsum.readthedocs.io/en/latest/path_finding.html#performance-comparison).

**Comment by mrader1248** [comment]
The problem with `einsum_path` from NumPy is that it's extremely slow.
```python
#  +---B---B---B---
#  |   |   |   |
#  A   |   |   |
#  |   |   |   |
#  +---B---B---B---
A = np.random.rand(16, 16)
B = np.random.rand(2, 16, 16)
ilabels = ["ab", "cad", "cbe", "fdg", "feh", "igj", "ihk"]
olabels = ilabels[-2][-1] + ilabels[-1][-1]
itensors = [A] + [B]*(len(ilabels)-1)
ishapes = [t.shape for t in itensors]
equ = ",".join(ilabels) + "->" + olabels

%timeit np.einsum_path(equ, *itensors, optimize=("optimal", np.iinfo(int).max))
596 ms ± 6.54 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%timeit einsum_optimize(ishapes, ilabels, olabels, optimize="dp")
8.48 ms ± 11.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```
With just 7 tensors (which is not really much for my field of research), my code is approximately 70 times faster. If I have 9 tensors using
```python
ilabels = ["ab", "cad", "cbe", "fdg", "feh", "igj", "ihk", "ljm", "lkn"]
```
my code needs 34.8 ms ± 1.18 ms and I cannot tell you, how long `einsum_path` needs, because I am not patient enough to wait for just a single run of it to finish.

**Comment by dgasmith** [comment]
`optimal` is a full tree search of every possible path which will ultimately give you the "best" path and you are also turning off sieving leading to very large timings. In general `optimal` is very naive and shouldn't be used except for challenging (and small!) problems. Using `optimal` defaults or the advanced `greedy` algorithm which has had quite a bit of effort into tuning and testing are all much faster than yours:
```
%timeit np.einsum_path(equ, *itensors, optimize=("optimal", np.iinfo(int).max))
664 ms ± 23.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
In [3]:

%timeit np.einsum_path(equ, *itensors, optimize="optimal")
491 µs ± 20.7 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
In [4]:

%timeit np.einsum_path(equ, *itensors, optimize="greedy")
243 µs ± 6.36 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

The quality of the path also matters a great deal in the overall cost of the contraction, I suggest reading into the `opt_einsum` repository for some exploratory work in performance for a variety of additional algorithms. I believe your current algorithm is identical to the "branch-1" algorithm in the `opt_einsum` docs.

In addition, it is likely argued that pure FLOP count is likely not very good for GPU implementations regardless and more advanced heuristics are likely needed, see [here](https://github.com/Bihaqo/tf_einsum_opt) for some comments on that.

**Comment by mrader1248** [comment]
No, my dynamical programming optimisation is not equivalent to your branch-1 algorithm. It searches for the optimal contraction order by first finding optimal contraction sequences for all pairs of tensors, then for all triples of tensors, and so on. Therefore, the solution is really optimal and still much faster than the full depth-first search. Maybe I should have given this more complex example already in the beginning:
```python
A = np.random.rand(16, 2, 16)
B = np.random.rand(2, 16, 16)
C = np.random.rand(2, 2, 2, 2)
ilabels = ["daf", "gebd", "ech", "ifk", "ljgi", "jhm", "abc"]
olabels = "klm"
itensors = [B,C,B,B,C,B,A]
ishapes = [t.shape for t in itensors]
equ = ",".join(ilabels) + "->" + olabels

np.einsum_path(equ, *itensors, optimize=("optimal", np.iinfo(int).max))[0][1:]
[(0, 6), (0, 5), (0, 4), (0, 3), (0, 2), (0, 1)]

einsum_optimize(ishapes, ilabels, olabels, optimize="dp")
[(0, 6), (0, 1), (0, 1), (0, 1), (0, 1), (0, 1)]
# note that einsum_path (einsum_optimize) appends (prepends) the
# contracted element to the list of tensors

np.einsum_path(equ, *itensors, optimize="optimal")[0][1:]
[(1, 4), (0, 1, 2, 3, 4, 5)]
# w/o knowledge about memory consumption this fails of course

np.einsum_path(equ, *itensors, optimize=("greedy", np.iinfo(int).max))[0][1:]
[(0, 3), (1, 3), (0, 1), (0, 1), (1, 2), (0, 1)]
# greedy does not find the optimal solution in this case

%timeit np.einsum_path(equ, *itensors, optimize=("optimal", np.iinfo(int).max))[0][1:]
612 ms ± 17.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%timeit einsum_optimize(ishapes, ilabels, olabels, optimize="dp")
11.1 ms ± 50.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%timeit np.einsum_path(equ, *itensors, optimize="optimal")[0][1:]
232 µs ± 11.1 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

%timeit np.einsum_path(equ, *itensors, optimize=("greedy", np.iinfo(int).max))[0][1:]
426 µs ± 11.2 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
Of course, `einsum_path` with strong memory restrictions is much faster as the search space is much smaller. But what does a fast solution help, if its result is basically useless?  Note that the previous example doesn't have tensors that are larger than the input tensors.

**Comment by dgasmith** [comment]
Saying its "basically useless" is quite disingenuous as there is a very large variety of use cases and expression types for `einsum` pathing (especially in the domain of very large numbers of tensors). In addition, the greedy approach finds a path that only costs 30% more than the most optimal approach while still being valid for virtually anything a user can give it. All algorithms will find different domains that they are "best" in which is why `opt_einsum` took the strategy of picking different algorithms depending on the problem input. 

For raw speed you might want to benchmark against the `opt_einsum` `optimal` implementation which seems a bit faster than yours:
```
%timeit oe.contract_path(equ, *itensors, optimize="optimal", memory_limit=np.iinfo(int).max)
8.93 ms ± 272 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

%timeit np.einsum_path(equ, *itensors, optimize=("optimal", np.iinfo(int).max))
659 ms ± 22.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

Hard to fully compare apples to apples here as I do not have your algorithms, but worth a look as the `np.einsum_path` times look comparable.

**Comment by mrader1248** [comment]
Sorry, I did not want to insult you. With useless I meant the solution `[(1, 4), (0, 1, 2, 3, 4, 5)]`, definitely not your overall work.
If you only have to perform a single contraction of this kind, 30 % more runtime seems to be an adequate option. However, if you have simulations running for weeks and months, repeating these contractions again and again, then you care about this extra runtime. Especially in the case of tensorflow, where the computation graph is built and optimised only once, but executed many times, a few milliseconds more of optimisation time are ok for me.

If you want to have a look into my implementation, you can find it [here](https://github.com/mrader1248/tensorflow/blob/0b024689d0e080abe459c6e0fa245b5d35af2d54/tensorflow/python/ops/special_math_ops.py#L336). Just copy the functions `einsum_optimize`, `_einsum_optimize_dp`, `_find_subgraphs`, `_einsum_optimize_dp_connected`, and `_tree_to_sequence`. You do not even need tensorflow to play around with it.

**Comment by martinwicke** [comment]
@MarkDaoust do you have additional concerns?

**Comment by mrader1248** [comment]
@martinwicke, @MarkDaoust: Anything else to do for me?

**Review by martinwicke — DISMISSED** [review]

**Comment by martinwicke** [comment]
@mrader1248 we'd be happy to review a PR that further improves optimizations.

**Comment by martinwicke** [comment]
However, this is much better than not having it, so we'll accept this now.

**Comment by rthadur** [comment]
@mrader1248 thank you for contribution , this change is failing sanity check , can you please check here once https://source.cloud.google.com/results/invocations/74054d17-9282-4740-a88c-c012f2bd67c4/targets/%2F%2Ftensorflow%2Ftools%2Fci_build:gen_ci_sanity_out/log

File: tensorflow/python/ops/special_math_ops.py [hunk 4]
```diff
@@ -272,38 +289,64 @@ def einsum(equation, *inputs, **kwargs):
...
+    missing_indices = set(input_axis_labels[0]) - set(output_axis_labels)
+
     if missing_indices:
       axis = [
-          i for i, a in enumerate(temp_axis_labels)
+          i for i, a in enumerate(input_axis_labels[0])
           if a not in output_axis_labels
       ]
-      temp = math_ops.reduce_sum(temp, axis=axis)
+      temp = math_ops.reduce_sum(inputs[0], axis=axis)
```
Comment (MarkDaoust) [inline tensorflow/python/ops/special_math_ops.py]: These `temp` variables are unused. I added a much more direct test for this. It fails at the "Invalid Equation" error below.

You know this code better than me, can you take a look?

**Review by MarkDaoust — CHANGES_REQUESTED** [review]
Hey @mrader1248, I'm trying to merge this.

I can run the auto-lint fixers and clear the lint errors, but there are some actual test failures. You can see the logs under:

Under [Checks ==> Linux GPU ==> //tensorflow/python:special_math_ops_test_gpu](https://source.cloud.google.com/results/invocations/64d1a1b1-e644-4dda-884b-e0c9b05e5393/targets/%2F%2Ftensorflow%2Fpython:special_math_ops_test_gpu/tests;group=__main__.EinsumTest;test=test_simple;row=2) :

They are:

```
test_ellipsis:
    ValueError: Invalid equation: ...ij,...jk->

test_simple:
    ValueError: Invalid equation: ijk,jklm->il
```

**Review by MarkDaoust — APPROVED** [review]

**Comment by MarkDaoust** [comment]
I got it fixed. You'll see the fix in the merge commit.

**Comment by MarkDaoust** [comment]
I tried to merge this, and found that there is equivalent functionality, based on https://github.com/dgasmith/opt_einsum currently being merged internally. After considering both options TensorFlow owners chose to go with the `opt_einsum`.

**Pull request closed without merging**